src/main/java/org/prebid/server/bidder/huaweiads/HuaweiAdmBuilder.java

```diff
@@ -0,0 +1,448 @@
+                        .map(Monitor::getUrlList)
+                        //todo: maybe addAll?
+                        .findFirst())
+                .map(list -> list.stream()
+                        .map(tracking -> "\"" + tracking + "\"")
+                        .collect(Collectors.joining(",")))
+                .orElse("");
+    }
+
+    private static String makeDspImpTrackings(Content content) {
+        return Optional.ofNullable(content.getMonitorList()).flatMap(list -> list.stream()
+                        .filter(monitor -> CollectionUtils.isNotEmpty(monitor.getUrlList())
+                                && MonitorEventType.of(monitor.getEventType()) == MonitorEventType.IMP)
+                        .map(Monitor::getUrlList)
+                        //todo: maybe addAll?
+                        .findFirst())
+                .map(list -> list.stream()
+                        .map(tracking -> "<img height=\"1\" width=\"1\" src='" + tracking + "' >  ")
+                        .collect(Collectors.joining()))
+                .orElse("");
+    }
+
+    public HuaweiAdm buildNative(AdsType adType, Content content, Native xNative) {
+        if (adType != AdsType.NATIVE) {
+            throw new PreBidException("extract Adm for Native ad: huaweiads response is not a native ad");
+        }
+
+        final MetaData metaData = content.getMetaData();
+        if (metaData == null) {
+            throw new PreBidException("Content.MetaData is empty");
+        }
+
+        final String clickUrl = getClickUrl(content);
+
+        final Request nativePayload = parseNativeRequest(xNative);
+        final List<Asset> assets = nativePayload.getAssets() == null
+                ? Collections.emptyList()
+                : nativePayload.getAssets();
+
+        Integer adHeight = null;
+        Integer adWidth = null;
+        final List<com.iab.openrtb.response.Asset> responseAssets = new ArrayList<>();
+
+        int imgIndex = 0;
+        int iconIndex = 0;
+        for (Asset asset : assets) {
+            final com.iab.openrtb.response.Asset.AssetBuilder responseAssetBuilder =
+                    com.iab.openrtb.response.Asset.builder();
+            if (asset.getTitle() != null) {
+                final String text = decode(metaData.getTitle());
+                final TitleObject titleObject = TitleObject.builder().text(text).len(text.length()).build();
+                responseAssetBuilder.title(titleObject);
+            } else if (asset.getVideo() != null) {
+                final HuaweiAdm admVideo = buildVideo(adType, content, null);
+                //todo: video always overwrites format
+                adHeight = admVideo.getHeight();
+                adWidth = admVideo.getWidth();
+                final VideoObject videoObject = VideoObject.builder().vasttag(admVideo.getAdm()).build();
+                responseAssetBuilder.video(videoObject);
+            } else if (asset.getImg() != null) {
```

Comment:
Please, extract each of `if else` branch into separated method.
Generally, try to refactor this code to next structure:
```
final boolean titleAsset = asset.getTitle() != null;
final boolean videoAsset = !titleAsset && asset.getVideo() != null;
final boolean imgAsset = !videoAsset && asset.getImg() != null;
final boolean dataAsset = !imgAsset && asset.getData() != null;

responseAsset = Asset.builder()
                .title(titleAsset ? titleAsset(...) : null)
                .video(...)
                ...
                .build()
```

Reply:
some global variables have to be updated within the assets iteration, but if the order of response assets could be not the same as the request assets order, it's possible to go over assets 4 times and resolve each type separately